src/libstore/build/substitution-goal.cc

```diff
@@ -140,6 +140,8 @@ void PathSubstitutionGoal::tryNext()
     {
         warn("substituter '%s' does not have a valid signature for path '%s'",
             sub->getUri(), worker.store.printStorePath(storePath));
+        warn("verify that your nix.conf contains a correct signature in 'trusted-public-keys' for %s",
+            sub->getUri());
```

Comment:
This message is not quite correct: `nix.conf` does not contain signatures but public keys; and the public keys are not *for* a specific substituter (so the `for %s` is misleading). I think it's better to combine the two warnings into something like:
```
warn("the substitute for '%s' from '%s' is not signed by any of the keys in 'trusted-public-keys'",
            worker.store.printStorePath(storePath), sub->getUri());
```
Or even better, we could show the key name from the signature (e.g. `'trusted-public-keys' does not contain a key for 'cache.nixos.org-1'`).

Reply:
This makes sense to me -- thanks for the swift review!

> Or even better, we could show the key name from the signature (e.g. 'trusted-public-keys' does not contain a key for 'cache.nixos.org-1').

Yeah this is great. Where are the possible places a user can/could have set `trusted-public-keys` out of curiosity? It might be good to say "This is where/how you can fix it" if you're starting from ground zero with nix.

Reply:
While I'm at this, I will try to fix this error too if that's alright.
Conversation from a Nix/NixOS channel in a developer slack a while ago

Swann = @evertedsphere

![image](https://user-images.githubusercontent.com/26604994/145422854-1f3c9bd3-332d-483b-a2ab-9730ab4b76a3.png)

Reply:
`trusted-public-keys` can be set in several configuration files (global, user's home directory), `flake.nix`, on the command line, and in an environment variable. So it's probably best not to refer to `nix.conf` specifically in the error message.